<h1>TOP 50 Spotify Tracks 2020</h1>
Data Processing with NumPy and Pandas project


<h2>Introduction</h2>
Welcome to Turing College Data Science program Sprint3 project. Project dedicated to practice working with data from Kaggle, performing basic Explanatory Data Analysis(EDA), performing queries and filtering data using Pandas. Objectives and requirements are provided in the README file.

__Structure of coding and analysis of this document__
1. Section 1. Preparing data
2. Section 2. EDA with 18 question
3. Section 3. Feature comparisson with different genres

EDA section provides questions and answers. There you can find most popular artist and albums. Filtering danceability and loudness scores. Longest and shortest tracks. Features correlation analysis.

In the end is summary.

<h3>Section 1. Preparing data</h3>

<h4>Load the data using Pandas</h4>
Importing csv, numpy and pandas libraries. Reading a csv file with 50 TOP Spotify tracks 2020.
First 5 rows provided from data.

In [1]:
import csv
import pandas as pd
import numpy as np

csv_path = "spotifytoptracks.csv"
df = pd.read_csv(csv_path, index_col=0)
df.head()

artist                                  album             track_name  \
0   The Weeknd                            After Hours        Blinding Lights   
1  Tones And I                           Dance Monkey           Dance Monkey   
2  Roddy Ricch  Please Excuse Me For Being Antisocial                The Box   
3    SAINt JHN                  Roses (Imanbek Remix)  Roses - Imanbek Remix   
4     Dua Lipa                       Future Nostalgia        Don't Start Now   

                 track_id  energy  danceability  key  loudness  acousticness  \
0  0VjIjW4GlUZAMYd2vXMi3b   0.730         0.514    1    -5.934       0.00146   
1  1rgnBhdG2JDFTbYkYRZAku   0.593         0.825    6    -6.401       0.68800   
2  0nbXyq5TXYPCO7pr3N8S4I   0.586         0.896   10    -6.687       0.10400   
3  2Wo6QQD1KMDWeFkkjLqwx5   0.721         0.785    8    -5.457       0.01490   
4  3PfIrDoz19wz7qK7tYeu62   0.793         0.793   11    -4.521       0.01230   

   speechiness  instrumentalness  liveness  valence    tempo  duration_ms  \
0       0.0598          0.000095    0.0897    0.334  171.005       200040   
1       0.0988          0.000161    0.1700    0.540   98.078       209755   
2       0.0559          0.000000    0.7900    0.642  116.971       196653   
3       0.0506          0.004320    0.2850    0.894  121.962       176219   
4       0.0830          0.000000    0.0951    0.679  123.950       183290   

               genre  
0           R&B/Soul  
1  Alternative/Indie  
2        Hip-Hop/Rap  
3   Dance/Electronic  
4           Nu-disco

<h4>Data cleaning.</h4>Missing values, removing duplicate samples and features, handling outliers.


<h5>Missing values</h5>
No missing values found so no further actions are needed.

In [2]:
df.isnull().sum() # Missing values count

artist              0
album               0
track_name          0
track_id            0
energy              0
danceability        0
key                 0
loudness            0
acousticness        0
speechiness         0
instrumentalness    0
liveness            0
valence             0
tempo               0
duration_ms         0
genre               0
dtype: int64

<h5>Duplicates</h5>
All rows doesn't have duplicates. No additional actions.

In [3]:
duplicates = df.duplicated()
sum_duplicates = duplicates.sum()
sum_duplicates = duplicates.size - duplicates.value_counts()[False]
print("Number of Duplicates: ", sum_duplicates)

Number of Duplicates:  0


<h5>Outliers</h5>
Detected 5 columns with outliers. Outliers values provided in code ouput below.
In regard this project scope and relatively small sample size outliers will remain in the data.

Columns with Outliers:
loudness
speechiness
instrumentalness
liveness
durati
ed:
49    312820
Name: duration_ms, dtype: int64

In [4]:
from scipy import stats

def find_outliers(column, method='zscore', threshold=3):
  """
  This function detects outliers in a numerical column using a specified method.

  Args:
      column (pandas.Series): The numerical column to analyze.
      method (str, optional): The method for outlier detection. Defaults to 'zscore'.
          - 'zscore': Uses z-scores (standard deviations from the mean).
          - 'iqr': Uses Interquartile Range (IQR).
      threshold (float, optional): The threshold for outlier detection. Defaults to 3.
          - For z-scores, values with absolute value exceeding the threshold are considered outliers.
          - For IQR, values below Q1 - threshold * IQR or above Q3 + threshold * IQR are considered outliers.

  Returns:
      pandas.Series: A Series containing True for outliers and False for non-outliers.
  """
  if method == 'zscore':
    z_scores = stats.zscore(column)
    return np.abs(z_scores) > threshold
  elif method == 'iqr':
    q1 = column.quantile(0.25)
    q3 = column.quantile(0.75)
    iqr = q3 - q1
    return (column < (q1 - threshold * iqr)) | (column > (q3 + threshold * iqr))
  else:
    raise ValueError("Invalid method. Choose 'zscore' or 'iqr'.")


numerical_columns = df.select_dtypes(include=['number'])

columns_with_outliers = []

for col in numerical_columns:
  # Detect outliers using z-scores
  outliers = find_outliers(df[col], method='zscore')

  # Track columns with outliers
  if outliers.any():
    columns_with_outliers.append(col)

  # Print results for the current column
  print(f"\nColumn: {col}")
  if outliers.any():
    print("Outliers detected:")
    print(df[col][outliers])  # Print rows with outliers in the current column
  else:
    print("No outliers detected.")

# Print summary of columns with outliers
if columns_with_outliers:
  print("\nSummary: Columns with Outliers:")
  for col in columns_with_outliers:
    print(col)
else:
  print("\nSummary: No outliers detected in any numerical columns.")


Column: energy
No outliers detected.

Column: danceability
No outliers detected.

Column: key
No outliers detected.

Column: loudness
Outliers detected:
24   -14.454
Name: loudness, dtype: float64

Column: acousticness
No outliers detected.

Column: speechiness
Outliers detected:
19    0.487
Name: speechiness, dtype: float64

Column: instrumentalness
Outliers detected:
24    0.657
Name: instrumentalness, dtype: float64

Column: liveness
Outliers detected:
2     0.790
41    0.792
Name: liveness, dtype: float64

Column: valence
No outliers detected.

Column: tempo
No outliers detected.

Column: duration_ms
Outliers detected:
49    312820
Name: duration_ms, dtype: int64

Summary: Columns with Outliers:
loudness
speechiness
instrumentalness
liveness
duration_ms


__<h3>Section 2. EDA with 21 question</h3>__
1. How many observations are there in this dataset?
2. 	How many features this dataset has
3. Which of the features are categorical?
4. Which of the features are numeric?
5. Are there any artists that have more than 1 popular track? If yes, which and how many?
6. Who was the most popular artist?
7. How many artists in total have their songs in the top 50?
8. Are there any albums that have more than 1 popular track? If yes, which and how many?
9. How many albums in total have their songs in the top 50?
10. Which tracks have a danceability score above 0.7?
11. Which tracks have a danceability score below 0.4?
12. Which tracks have their loudnes above -5?
13. Which tracks have their loudness below -8?
14. Which track is the longest?
15. Which track is the shortest?
16. Which genre is the most popular?
17. Which genres have just one song on the top 50?
18. How many genres in total are represented in the top 50?
19. Which features are strongly positively correlated?
20. Which features are strongly negatively correlated?
21. Which features are not correlated?


 are not correlated?

1. How many observations are there in this dataset?

In [5]:
# 1. How many observations are there in this dataset?
df.count()
row_count = len(df)
print(f'The DataFrame has {row_count} rows.')

The DataFrame has 50 rows.


2. How many features this dataset has?

In [6]:
#2	How many features this dataset has?
feature_count = len(df.columns)
print(f"This dataset has {feature_count} features")


This dataset has 16 features


3. Which of the features are categorical?

In [7]:
# 3. Which of the features are categorical?

categorical_features = [col for col in df.columns if df[col].dtypes == object]
print("Categorical features:", categorical_features)


Categorical features: ['artist', 'album', 'track_name', 'track_id', 'genre']


4. Which of the features are numeric?

In [8]:
# 4. Which of the features are numeric?
# Select columns with numeric data types
numerical_features = df.select_dtypes(include=[np.number])  # Import numpy as np

print("Numerical Features:", numerical_features.columns.tolist())


Numerical Features: ['energy', 'danceability', 'key', 'loudness', 'acousticness', 'speechiness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'duration_ms']


5. Are there any artists that have more than 1 popular track? If yes, which and how many?

In [9]:
# 5. Are there any artists that have more than 1 popular track? If yes, which and how many?
count_popular_artists = df['artist'].value_counts()
print(count_popular_artists[count_popular_artists > 1])  


artist
Billie Eilish    3
Dua Lipa         3
Travis Scott     3
Justin Bieber    2
Harry Styles     2
Lewis Capaldi    2
Post Malone      2
Name: count, dtype: int64


6. Who was the most popular artist?

In [10]:
# 6. Who was the most popular artist?
most_popular_artists = count_popular_artists[count_popular_artists == count_popular_artists.max()]
print("Most popular artists:\n", most_popular_artists)

Most popular artists:
 artist
Billie Eilish    3
Dua Lipa         3
Travis Scott     3
Name: count, dtype: int64


7. How many artists in total have their songs in the top 50?

In [11]:
# 7. How many artists in total have their songs in the top 50?
#counting unique artists names
total_artists_in_top50 = count_popular_artists.index.nunique()

print("Total number of artists who have their songs in the top 50 is: ", total_artists_in_top50)


Total number of artists who have their songs in the top 50 is:  40


8. Are there any albums that have more than 1 popular track? If yes, which and how many?

In [12]:
# 8. Are there any albums that have more than 1 popular track? If yes, which and how many?
albums = df['album'].value_counts()
albums_with_more_than1_track = albums[albums > 1]
print("Albums that have more than 1 popular track: \n", albums_with_more_than1_track)


Albums that have more than 1 popular track: 
 album
Future Nostalgia        3
Hollywood's Bleeding    2
Fine Line               2
Changes                 2
Name: count, dtype: int64


9. How many albums in total have their songs in the top 50?

In [13]:
# 9. How many albums in total have their songs in the top 50?
total_albums_in_top50 = albums.index.nunique()
print("Total albums in top 50: \n", total_albums_in_top50)

Total albums in top 50: 
 45


10. Which tracks have a danceability score above 0.7?

In [37]:
# 10. Which tracks have a danceability score above 0.7?
print("Tracks with danceability above 0.7:")
list = ['track_name', 'danceability']
df[df['danceability'] > 0.7][list].sort_values(by='danceability', ascending=False)

Tracks with danceability above 0.7:


track_name  danceability
27                WAP (feat. Megan Thee Stallion)         0.935
2                                         The Box         0.896
39                                        Ride It         0.880
28                                    Sunday Best         0.878
33               Supalonely (feat. Gus Dapperton)         0.862
40                                     goosebumps         0.841
49                                     SICKO MODE         0.834
15                                   Toosie Slide         0.830
1                                    Dance Monkey         0.825
29                    Godzilla (feat. Juice WRLD)         0.808
14                       Intentions (feat. Quavo)         0.806
10                                           Tusa         0.803
19                     Life Is Good (feat. Drake)         0.795
4                                 Don't Start Now         0.793
22                                    Breaking Me         0.789
17                                         Say So         0.787
3                           Roses - Imanbek Remix         0.785
8                                         Falling         0.784
38                                          Hawái         0.783
13                                Blueberry Faygo         0.774
20               Savage Love (Laxed - Siren Beat)         0.767
18                                       Memories         0.764
25                                       Señorita         0.759
37  Sunflower - Spider-Man: Into the Spider-Verse         0.755
32                                       Dynamite         0.746
5                    ROCKSTAR (feat. Roddy Ricch)         0.746
31                                 Break My Heart         0.730
7                death bed (coffee for your head)         0.726
41                      RITMO (Bad Boys For Life)         0.723
42                                     THE SCOTTS         0.716
24                            everything i wanted         0.704
26                                        bad guy         0.701

11. Which tracks have a danceability score below 0.4?

In [38]:
# 11. Which tracks have a danceability score below 0.4?
print("Tracks with danceability below 0.4:")
list = ['track_name', 'danceability']
df[df['danceability'] < 0.4][list]

Tracks with danceability below 0.4:


track_name  danceability
44  lovely (with Khalid)         0.351

12. Which tracks have their loudness above -5?

In [41]:
# 12. Which tracks have their loudness above -5?
print("Tracks with loudness above -5:")
list = ['track_name', 'loudness']
df[df['loudness'] > -5][list].sort_values(by='loudness', ascending=False)

Tracks with loudness above -5:


track_name  loudness
10                                           Tusa    -3.280
40                                     goosebumps    -3.370
31                                 Break My Heart    -3.434
38                                          Hawái    -3.454
12                                        Circles    -3.497
23                         Mood (feat. iann dior)    -3.558
21                                      Adore You    -3.675
49                                     SICKO MODE    -3.714
48                                       Physical    -3.756
35                Rain On Me (with Ariana Grande)    -3.764
43                                        Safaera    -4.074
6                                Watermelon Sugar    -4.209
39                                        Ride It    -4.258
37  Sunflower - Spider-Man: Into the Spider-Verse    -4.368
32                                       Dynamite    -4.410
4                                 Don't Start Now    -4.521
17                                         Say So    -4.577
33               Supalonely (feat. Gus Dapperton)    -4.746
16                                  Before You Go    -4.858

13. Which tracks have their loudness below -8?

In [44]:
# 13. Which tracks have their loudness below -8?
print("Tracks with loudness below -8:")
list = ['track_name', 'loudness']
df[df['loudness'] < -8][list].sort_values(by='loudness', ascending=False)


Tracks with loudness below -8:


track_name  loudness
20                Savage Love (Laxed - Siren Beat)    -8.520
8                                          Falling    -8.756
36                             HIGHEST IN THE ROOM    -8.764
7                 death bed (coffee for your head)    -8.765
15                                    Toosie Slide    -8.820
47  If the World Was Ending - feat. Julia Michaels   -10.086
44                            lovely (with Khalid)   -10.109
26                                         bad guy   -10.965
24                             everything i wanted   -14.454

14. Which track is the longest?

In [18]:
# 14. Which track is the longest?
max_duration_index = df['duration_ms'].idxmax()
longest_track = df.loc[max_duration_index, 'track_name']
print("Longest track:")
longest_track

Longest track:


'SICKO MODE'

15. Which track is the shortest?

In [19]:
# 15. Which track is the shortest?
min_duration_index = df['duration_ms'].idxmin()
shorthest_track = df.loc[min_duration_index, 'track_name']
print("Shortest track:")
shorthest_track


Shortest track:


'Mood (feat. iann dior)'

16. Which genre is the most popular?

In [20]:
# 16. Which genre is the most popular?
count_genre = df['genre'].value_counts()
most_popular_genre = count_genre[count_genre == count_genre.max()]
print("Most popular genre:\n", most_popular_genre)

Most popular genre:
 genre
Pop    14
Name: count, dtype: int64


17. Which genres have just one song on the top 50?

In [21]:
# 17. Which genres have just one song on the top 50?
genre_with_one_song = count_genre[count_genre == 1]
print("Genre in top 50 with only one song:\n", genre_with_one_song.index.tolist())

Genre in top 50 with only one song:
 ['Nu-disco', 'R&B/Hip-Hop alternative', 'Pop/Soft Rock', 'Pop rap', 'Hip-Hop/Trap', 'Dance-pop/Disco', 'Disco-pop', 'Dreampop/Hip-Hop/R&B', 'Alternative/reggaeton/experimental', 'Chamber pop']


18. How many genres in total are represented in the top 50?

In [22]:
# 18. How many genres in total are represented in the top 50?
total_genres_in_top50 = count_genre.index.nunique()
print("In the top 50 total genres: ", total_genres_in_top50)



In the top 50 total genres:  16


19. Which features are strongly positively correlated?


In [23]:
# 19. Which features are strongly positively correlated?
treshold = 0.7
numeric_columns = df.select_dtypes(include=['number'])
matrix = numeric_columns.corr()
#matrix.head()
strong_positive_correlations = matrix.where(~np.tril(np.ones(matrix.shape)).astype(bool))
strong_positive_correlations = strong_positive_correlations[strong_positive_correlations > treshold]

print("Strongly positively correlated numeric features in TOP 50 Spotify Tracks 2020")
# print in a formatted table excluding NaN values that do not met selected treshold
strong_positive_correlations.style.format(na_rep='')  



Strongly positively correlated numeric features in TOP 50 Spotify Tracks 2020


20. Which features are strongly negatively correlated?

In [24]:
# 20. Which features are strongly negatively correlated?
treshold = -0.6
numeric_columns = df.select_dtypes(include=['number'])
matrix = numeric_columns.corr()
matrix.head()
strong_negative_correlations = matrix.where(~np.tril(np.ones(matrix.shape)).astype(bool))
strong_negative_correlations = strong_negative_correlations[strong_negative_correlations < treshold]

print("Strongly negatively correlated numeric features in TOP 50 Spotify Tracks 2020")
# print in a formatted table excluding NaN values that do not met selected treshold
strong_negative_correlations.style.format(na_rep='')  

Strongly negatively correlated numeric features in TOP 50 Spotify Tracks 2020


21. Which features are not correlated?

**Negatively correlated features.** Treshold for negatively correlated features choosen to be valid for moderately strong correlations. Base line for lower than -0.7 treshold doesn't apply for this data.

**Not Correlated features.** Selected treshold (-0.1:0.1) results in many features with no correlation. Although correlation is near 0 it doesn't necessarily mean there's no relationship between variables. Near zero correlation indicates there are no linear relationship between some features. Further study needs to be done to properly evalute statistical significance and possible realtionships.

In [25]:
# 21. Which features are not correlated?
treshold_lower = -0.1
treshold_upper = 0.1
numeric_columns = df.select_dtypes(include=['number'])
matrix = numeric_columns.corr()
matrix.head()
no_correlations = matrix.where(~np.tril(np.ones(matrix.shape)).astype(bool))
no_correlations = no_correlations[(no_correlations < treshold_upper) & (no_correlations > treshold_lower) ]

print("Not correlated numeric features in TOP 50 Spotify Tracks 2020")
# print in a formatted table excluding NaN values that do not met selected treshold
no_correlations.style.format(na_rep='')  

Not correlated numeric features in TOP 50 Spotify Tracks 2020


<h3>Section 3. Feature comparisson with different genres</h3>

How does the Danceability score compare between Pop, Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie genres?

In [26]:
#How does the danceability score compare between Pop, Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie genres?
genre_names = ['Pop', 'Hip-Hop/Rap', 'Dance/Electronic', 'Alternative/Indie']
feature_name = 'danceability'

filtered_df = df[df['genre'].isin(genre_names)]
genre_stats = filtered_df.groupby('genre')[feature_name].describe()
print("Danceability comparisson main statistics")
genre_stats.style.format(precision=2)

Danceability comparisson main statistics


**Danceability comparisson**

Danceability distribution between Pop, Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie genres are in relatively small range (0.66 - 0.78). This suggest that there might not be a signicant difference in overall user preference between these genres. 
Hip-Hop/Rap adm Dance/Electronic has highest average ratings(0.765 and 0.755) in danceablity. This could indicate a slight preference to dance music choices. However it is a small difference in averages and not enough data to draw strong conclusions. It's important to consider the limitations of the data. With only 4 and 5 samples for Alternative/Indie and Dance/Electronic respectively conclusions may not be correctly evaluated.

How does the loudness score compare between Pop, Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie genres?

In [27]:
#How does the loudness score compare between Pop, Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie genres?
genre_names = ['Pop', 'Hip-Hop/Rap', 'Dance/Electronic', 'Alternative/Indie']
feature_name = 'loudness'

filtered_df = df[df['genre'].isin(genre_names)]
genre_stats = filtered_df.groupby('genre')[feature_name].describe()
print("Loudness comparisson main statistics")
genre_stats.style.format(precision=2)

Loudness comparisson main statistics


**Loudness comparisson**

Loudness measured in decibels(dB) in negative values.
Important to consider the sample size. 2 of 4 features has 5 or less features and remaining has 13 and 14. Comparing max values there are similar values for all genres. Altough the lowest min value is around double lower for POP genre. For correct approach deeper investigation in the usage of dB or dBSPL(Sound pressure Lever).

How does the acousticness score compare between Pop, Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie genres?

In [28]:
#How does the acousticness score compare between Pop, Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie genres?
genre_names = ['Pop', 'Hip-Hop/Rap', 'Dance/Electronic', 'Alternative/Indie']
feature_name = 'acousticness'

filtered_df = df[df['genre'].isin(genre_names)]
genre_stats = filtered_df.groupby('genre')[feature_name].describe()
print("Acousticness comparisson main statistics")
genre_stats.style.format(precision=2)

Acousticness comparisson main statistics


**Acousticness comparisson**

"Mean" values is in range (0.1 to.58) suggest different acousticness properties. Higher values for Alternative/Indie genre might indicate more aquistic instruments for example guitars. Lower values as for Dance/Electronic might have more drum machines and synthesizers.

"STD"(Standard Deviation) lower for Dance/Electronic(0.10) indicating songs to have average acousticness for the genre. Higher deviation for POP(0.32) might show variation between 14 provided songs in the sample. There might be subgenres in the POP genre with some songs having more acoustic elements than others.


<h3>Summary</h3>

__Insights__
1. Data set has 16 features and 50 rows.
2. Data set contains numerical and object data value types.
3. Most popular artists are Billy Eilish, Dua Lipa and Travis Scoot.
4. Most popular album in Spotify TOP 50 track of 2020 was 'Future notalife'
5. Longest song 'SICKO ME', shortest song was 'Mood(feat. iann der)
6. Highest positive correlation 0.79 is between loudness and energy. More sound equals more energy.
7. Highest negative correlation -0.68 is between anergy and acousticness. More different types of instruments in songstends to lower energy levels.
8. There are many not correlated features with correlation in (-0.1-0.1). It shows that features doesn't have linear relationship. Further investigation needed about possible relations.
9. Comparisson requirements requires bigger data samples for drawing clear conclusions. Sample size of 4, 5, 13 and 14 seems to be not sufficient for correct results. Loudness feature requeres additional investigation in the meaning behind negative values and industry practices.


__Suggestions for improvements__
1. Bigger data sample.
2. Include graphical materials for visualisation(Boxplots, histograms).
3. Improve EDA specific questions with more broad analysis more focusing on insightful invvestigation to provide for the product manager that does actual make hit on Spotify. This might also iclude exploring songs that did not actually got to the TOP 50 on 2020.